File: Sources/HaviKit/Transport/HaviSubmitResult.swift
import Foundation

/// How the capture sheet should present a failed submit (design §2, §4). Auth /
/// workspace failures show an actionable *Reconnect HAVI*; transient failures a
/// *Retry*; terminal failures dismiss with the reason and no retry.
public enum HaviSubmitFailureKind: Sendable, Equatable {
    case retry
    case reconnect
    case terminal
}

public struct HaviSubmitFailure: Sendable, Equatable {
    public let userMessage: String
    public let kind: HaviSubmitFailureKind
    public let code: String?

    public init(userMessage: String, kind: HaviSubmitFailureKind, code: String?) {
        self.userMessage = userMessage
        self.kind = kind
        self.code = code
    }
}

/// Result of a best-effort foreground submit. There is no on-disk outbox
/// (design §4) — the sheet stays open on `.failure`, drawing + comment intact.
public enum HaviSubmitResult: Sendable, Equatable {
    case success(id: String?)
    case failure(HaviSubmitFailure)
}

File: android/havikit-core/src/main/kotlin/ai/handgemacht/havikit/HaviSubmitResult.kt
package ai.handgemacht.havikit

/**
 * How the capture sheet should present a failed submit (wire spec §7). Auth /
 * workspace failures show an actionable *Reconnect HAVI*; transient failures a
 * *Retry*; terminal failures dismiss with the reason and no retry.
 */
public enum class HaviSubmitFailureKind {
    RETRY,
    RECONNECT,
    TERMINAL,
}

public data class HaviSubmitFailure(
    val userMessage: String,
    val kind: HaviSubmitFailureKind,
    val code: String?,
)

/**
 * Result of a best-effort foreground submit. There is no on-disk outbox — the
 * sheet stays open on [Failure], drawing + comment intact (wire spec §5.2, §7).
 */
public sealed interface HaviSubmitResult {
    public data class Success(val id: String?) : HaviSubmitResult

    public data class Failure(val failure: HaviSubmitFailure) : HaviSubmitResult
}
